Add Newton method

vladimir-ch opened the pull request:
Has the time really come for Newton?

An implementation of a pure, locally convergent Newton method without linesearch and of a modified, globally convergent Newton method with a linesearch.

The pure Newton is rather simple now that mat64 has symmetric matrices and Cholesky decomposition.

The modification in ModifiedNewton is a simple one, just add multiples of identity to the Hessian until it becomes positive definite. I did not venture into the modified Cholesky method at the moment but I can if there is a desire to have it.

I added three functions for resizing/reusing matrices in a similar way we already resize/reuse slices. Experts on matrix package @kortschak @btracey might have a better suggestion.

No tests at the moment, first we need to add Hess() method to functions, then modify the testing itself to accomodate Hessian-based methods.

PTAL, @btracey

kortschak commented on types.go:
```diff
@@ -257,3 +257,24 @@ func resize(x []float64, dim int) []float64 {
 	return x[:dim]
 }
+
+func resizeDense(m *mat64.Dense, r, c int) *mat64.Dense {
```
Use Dense.View for this.

The other types don't have that though, so leave them until we support that.

Note @btracey, this use of View is not supported as well as it used to be because View returns a Matrix(*Dense) and this could all be done in place.

kortschak commented on newton.go:
```diff
@@ -0,0 +1,197 @@
+		// Modified Hessian is not PD, so increase tau.
+		n.tau = math.Max(n.Increase*n.tau, 0.001)
+	}
+	d := mat64.NewDense(dim, 1, dir)
```
Add a comment that this is operating via side effect on dir's backing array.

vladimir-ch commented on newton.go:
```diff
@@ -0,0 +1,203 @@
+	dim := len(loc.X)
+	n.chol = resizeTriDense(n.chol, dim)
+	r, _ := n.step.Dims()
+	if dim < r {
```
Do you mean using Grow and View like this, @kortschak ?

kortschak replied:
Yes, that looks good.

btracey commented on newton.go:
```diff
@@ -0,0 +1,203 @@
+// For a globally convergent Hessian-based method, see ModifiedNewton. If the
+// Hessian matrix cannot be formed explicitly or if the computational cost of
+// its factorization is prohibitive, BFGS or L-BFGS methods can be used
+// instead.
```
I'm not sure this should be in this package. The package is for general nonlinear minimization, and it is very easy for a general nonlinear function to have a non-positive-definite hessian, or to diverge in the iterative Newton procedure. This algorithm is useful somewhere, but I would rather have the Newton method in this package be one that can handle the general class of functions.

vladimir-ch replied:
I wasn't sure either. I included it because the method is very simple, just the Hessian and its factorization, no line search, and the docs say that it can fail if the starting point is far from a minimum. I have no problem removing it.

btracey commented on newton.go:
```diff
@@ -0,0 +1,203 @@
+
+// ModifiedNewton implements a modified Newton's method for unconstrained
+// minimization of functions that satisfy Gradient and Hessian interfaces. It
+// is a globally convergent line-search method that can be applied from any
```
We need to find an alternate expression to 'globally convergent'. That implies it will find a global minimum.

// Newton implements a modified Newton's method for Hessian-based 
// unconstrained minimization. It applies regularization when the 
// Hessian is non-positive-definite, and uses line-searches so it can
// converge to a local minimum from any starting point.

vladimir-ch replied:
I think that "globally convergent method" is a standard terminology for a method that converges from any starting point, not that it can find a global minimum. I agree that it can be confusing, though. It is probably best to avoid this term.

btracey commented on newton.go:
```diff
@@ -0,0 +1,203 @@
+			break
+		}
+		// Modified Hessian is not PD, so increase tau.
+		n.tau = math.Max(n.Increase*n.tau, 0.001)
```
Why is there a maximum here?

btracey replied:
This doesn't actually increase tau if n.tau == 0

vladimir-ch replied:
If n.tau == 0, it does increase to 0.001.

btracey replied:
Sorry, I'm an idiot.

vladimir-ch replied:
I don't think so.

btracey commented on newton.go:
```diff
@@ -0,0 +1,203 @@
+		// Hessian first.
+		n.tau = 0
+	} else if n.tau == 0 {
+		// Use the last tau that gave PD modified Hessian.
```
I don't understand the logic. If all the elements are positive, try the normal solve. Agreed. If they aren't all positive, only modify tau if the last solve went fine? What about the case where minA <= 0 and n.tau != 0?

vladimir-ch replied:
n.tau == 0 if it is the first iteration or if on the previous iteration no modification of the Hessian was necessary. In that case, if minA < 0, modify tau. If the Hessian was modified on the previous iteration, assume that a similar modification will be necessary at the current iteration and so reuse n.tau directly instead of starting from -minA + 0.001.

The value 0.001 is taken from Nocedal&Wright. I can make it a settable parameter, but I don't know how to name and document it succinctly and what the recommendation for its values should be.

vladimir-ch replied:
Well, in fact, I would know how to name and document it, it is just not a terribly important parameter and the number of factorizations can already be controlled with ModifiedNewton.Increase.

btracey replied:
Where is this in N&W? My copy suggests using the Frobenius norm of the matrix to set tau.

btracey replied:
Oh! The comment is misleading. The comment in the block applies to the missing "else", not to the code right there. I think it's better to have a comment above the block. Something like
````
// If the smallest diagonal entry is positive, the Hessian may be positive definite, and so first attempt an un-modified cholesky. If the smallest entry is negative, use the final tau from the last iteration if regularization was needed, otherwise guess an appropriate value for tau
````

vladimir-ch replied:
Sorry for the confusion, I didn't give enough thought to where to put the comment.

Modified as you suggested.

btracey commented on newton.go:
```diff
@@ -0,0 +1,143 @@
+
+// Newton implements a modified Newton's method for Hessian-based unconstrained
+// minimization. It applies regularization when the Hessian is not positive
+// definite, and uses line-searches so it can converge to a local minimum from
```
I don't think the "uses line searches" is needed here. It's described below, and isn't relevant to the important information (the rest of the paragraph)

vladimir-ch replied:
I'm not completely sure about this, but I think that for the theory a line search (together with some kind of Hessian modification) is necessary to get convergence from any starting point. The wording tried to reflect that.

I removed "uses line-searches so".

btracey commented on newton.go:
```diff
@@ -0,0 +1,143 @@
+// a line search.
+//
+// Away from a minimizer H_k may not be positive definite and d_k may not be a
+// descent direction. Newton implements a Hessian modification strategy that
```
that adds successively

vladimir-ch replied:
Done

btracey commented on newton.go:
```diff
@@ -0,0 +1,143 @@
+// can be used instead.
+type Newton struct {
+	// LinesearchMethod is a method used for selecting suitable steps along the
+	// descent direction d. Steps should satisfy the Wolfe, Goldstein or Armijo
```
should satisfy at least one of the

vladimir-ch replied:
Done

btracey commented on newton.go:
```diff
@@ -0,0 +1,143 @@
+	// conditions. If LinesearchMethod == nil, an appropriate default is
+	// chosen.
+	LinesearchMethod LinesearchMethod
+	// Increase is a factor by which a scalar tau is successively increased so
```
Increase is the factor by which

vladimir-ch replied:
Done

btracey commented on newton.go:
```diff
@@ -0,0 +1,143 @@
+	n.chol = resizeTriDense(n.chol, dim)
+	n.hess = resizeSymDense(n.hess, dim)
+	n.tau = 0
+	n.computeNextDir(loc, dir)
```
Why does this need its own method and not just call NextDirection directly?

vladimir-ch replied:
Errrrrr ... you are right, of course. Fixed.

btracey commented on newton.go:
```diff
@@ -0,0 +1,143 @@
+	return 1
+}
+
+func (n *Newton) computeNextDir(loc *Location, dir []float64) {
```
We should cite Nocedal and Wright here since that's where you got the algorithm from.

vladimir-ch replied:
Done

btracey commented on newton.go:
```diff
@@ -0,0 +1,143 @@
+	linesearch *Linesearch
+
+	hess *mat64.SymDense
+	chol *mat64.TriDense
```
Comment that this is storage for the cholesky.

vladimir-ch replied:
Done

btracey commented on newton.go:
```diff
@@ -0,0 +1,143 @@
+		n.tau = math.Max(n.Increase*n.tau, 0.001)
+	}
+	d := mat64.NewDense(dim, 1, dir)
+	// Store the solution in d's backing array, dir.
```
Add a TODO that this should use mat64.Vector SolveCholesky when it exists (or, if you're feeling charitable, implement it :) ). This way we can use Dtrsv instead of Dtrsm

vladimir-ch replied:
Done

btracey commented on newton.go:
```diff
@@ -0,0 +1,143 @@
+				n.hess.SetSym(i, i, loc.Hessian.At(i, i)+n.tau)
+			}
+		}
+		// Try to apply the Cholesky decomposition.
```
We should use the Upper cholesky here instead of the lower. I'm pretty sure it doesn't matter as far as the algorithm is concerned (since we're just using it to perform a solve next). However, SymDense is assumed to be UpperTriangular, and the eventual copy into the triangular is going to be faster into an upper triangular than a lower triangular.

vladimir-ch replied:
Done

btracey commented on types.go:
```diff
@@ -257,3 +257,17 @@ func resize(x []float64, dim int) []float64 {
+
+func resizeTriDense(m *mat64.TriDense, dim int) *mat64.TriDense {
+	if m == nil || cap(m.RawTriangular().Data) < dim*dim {
+		return mat64.NewTriDense(dim, false, nil)
```
See above, I think this should be upper.

vladimir-ch replied:
Done

vladimir-ch commented on newton.go:
```diff
@@ -0,0 +1,145 @@
+		n.tau = -minA + 0.001
+	}
+
+	for {
```
I am not sure about this infinite loop. If `loc.Hessian` is reasonable, it will eventually break at line 122. But what if it contains some garbage? I can put a limit on the number of iterations here, but since NextDirection() does not return an error, what would be an appropriate response if the limit is reached? Just return the negative gradient? Comments, @btracey @kortschak ?

btracey replied:
Yea, we need to do something. If the Hessian has an "inf" for example it will never conclude. Trying a few times and then doing the negative gradient seems reasonable. The other option is to return an error, but most users would probably rather find the optimum.

vladimir-ch replied:
Ok, added `const maxNewtonModifications = 20`. The number is just a shot from the hip.

btracey commented on newton.go:
```diff
@@ -0,0 +1,145 @@
+
+	d := mat64.NewDense(dim, 1, dir)
+	// Store the solution in d's backing array, dir.
+	// TODO(vladimir-ch): This should use mat64.Vector.SolveCholesky() when it exists.
```
Why is this a TODO? We have a Vector.SolveCholesky

:-)

vladimir-ch replied:
Oooh, you are so fast :-)

vladimir-ch commented:
Addressed all the issues and newly also added tests for Newton. Failing Travis build is again something with Travis, the tests all pass. PTAL, @btracey

btracey commented:
LGTM

vladimir-ch commented:
I can't believe it, the Newton odyssey is over! And originally I thought it would be just a couple of commits. Thanks for bearing with me @btracey @kortschak .

Ta ta taaa (the one from The Croods)
